src/otx/algo/detection/utils/yolov7_v9_utils.py

```diff
@@ -0,0 +1,256 @@
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Utils for YOLOv7 and v9.
+
+Reference : https://github.com/WongKinYiu/YOLO
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+import torch
+from einops import rearrange
+from torch import Tensor, nn
+from torchvision.ops import batched_nms
+
+if TYPE_CHECKING:
+    from otx.algo.detection.detectors import SingleStageDetector
+
+
+def auto_pad(kernel_size: int | tuple[int, int], dilation: int | tuple[int, int] = 1, **kwargs) -> tuple[int, int]:  # noqa: ARG001
+    """Auto Padding for the convolution blocks.
+
+    Args:
+        kernel_size (int | tuple[int, int]): The kernel size of the convolution block.
+        dilation (int | tuple[int, int]): The dilation rate of the convolution block. Defaults to 1.
+
+    Returns:
+        tuple[int, int]: The padding size for the convolution block.
+    """
+    if isinstance(kernel_size, int):
+        kernel_size = (kernel_size, kernel_size)
+    if isinstance(dilation, int):
+        dilation = (dilation, dilation)
+
+    pad_h = ((kernel_size[0] - 1) * dilation[0]) // 2
+    pad_w = ((kernel_size[1] - 1) * dilation[1]) // 2
+    return pad_h, pad_w
+
+
+def round_up(x: int | Tensor, div: int = 1) -> int | Tensor:
+    """Rounds up `x` to the bigger-nearest multiple of `div`.
+
+    Args:
+        x (int | Tensor): The input value.
+        div (int): The divisor value. Defaults to 1.
+
+    Returns:
+        int | Tensor: The rounded up value.
+    """
+    return x + (-x % div)
+
+
+def generate_anchors(image_size: tuple[int, int], strides: list[int]) -> tuple[Tensor, Tensor]:
+    """Find the anchor maps for each height and width.
+
+    Args:
+        image_size (tuple[int, int]): the image size of augmented image size.
+        strides list[int]: the stride size for each predicted layer.
+
+    Returns:
+        tuple[Tensor, Tensor]: The anchor maps with (HW x 2) and the scaler maps with (HW,).
+    """
+    height, width = image_size
+    anchors = []
+    scaler = []
+    for stride in strides:
+        anchor_num = width // stride * height // stride
+        scaler.append(torch.full((anchor_num,), stride))
+        shift = stride // 2
+        h = torch.arange(0, height, stride) + shift
+        w = torch.arange(0, width, stride) + shift
+        anchor_h, anchor_w = torch.meshgrid(h, w, indexing="ij")
+        anchor = torch.stack([anchor_w.flatten(), anchor_h.flatten()], dim=-1)
+        anchors.append(anchor)
+    all_anchors = torch.cat(anchors, dim=0)
+    all_scalers = torch.cat(scaler, dim=0)
+    return all_anchors, all_scalers
+
+
+def transform_bbox(bbox: Tensor, indicator: str = "xywh -> xyxy") -> Tensor:
```

Comment:
Can't we use torchvision functionality to convert bboxes as we did for RTDetr, for example?

Reply:
As I checked, this module is for yolov7. So I removed at bddc817 for now and I'll consider your suggestion for yolov7 integration.